### ISSUE TITLE
c++ benchmarks

### ISSUE THREAD

**Issue by jianshu93** [frjnn/bhtsne#10]
Dear Francesco,

Thanks for this implementation. I assume it is faster than the C++ original version right considering the rayon parallel efficiency. I will have some tests later on but just want to know whether you have some benchmarks.

Thanks,

Jianshu

**Comment by frjnn** [comment]
Don't have any benchmarks, your effort is very welcome.

**Comment by jianshu93** [comment]
Hello frjnn,

See the error I have (I created an examples folder and put the example code in markdown in to it and then run cargo run --example iris.rs):

Compiling bhtsne v0.5.0 (/Users/jianshuzhao/Github/bhtsne-rs)
error: expected item, found keyword `let`
  --> examples/mnist.rs:17:1
   |
17 | let data: Vec<f32> = bhtsne::load_csv("iris.csv", true, Some(&[4]), |float| {
   | ^^^ expected item

error: could not compile `bhtsne` due to previous error

[iris.rs.zip](https://github.com/frjnn/bhtsne/files/7673782/iris.rs.zip)


Sorry I am new to rust..

**Comment by frjnn** [comment]
No problem, I set up a small program for you, don't forget to run in release mode.

[benches.zip](https://github.com/frjnn/bhtsne/files/7674474/benches.zip)

**Comment by jianshu93** [comment]
Hello,

It works this time for iris:

   Compiling bhtsne v0.5.0 (/Users/jianshuzhao/Github/bhtsne-rs)
    Finished dev [unoptimized + debuginfo] target(s) in 0.36s
     Running `target/debug/examples/main`
Mean avg. time over 10 trials = 386.4 ms

This is faster than the C++ version


But for the mnist (70,000 lines,784 dimentions, no dimension reduction), it is very slow and takes more than 30 minutes or so (still running). And I notice that no more than 1 thread was actually used when running it. The memory requirement bumps up to 60G, which is surprising. And idea?

there is no sample name in the embedding csv output (row name of original csv input but I assume it keeps the original order of samples). Would it be easy to add sample name in the output as the row name?

Thanks,

Jianshu

**Comment by jianshu93** [comment]
The distance calculation is not paralleled (thus very slow) while tsne itself is paralleled. I am thinking whether we should add some fast distance computation module.

Thanks,

Jianshu

**Comment by frjnn** [comment]
> I assume it keeps the original order of samples.

Correct.

> Can we use this SIMD (...)

SIMD will be officially supported by the Rust standard library in the (somewhat) near future, I'd prefer to wait until then to make a move in that direction.

Concerning what you have pointed out:

* **memory usage** The algorithm uses logically padded matrices to avoid false sharing and squeeze every bit of performance.
* **unreduced MNIST Barnes Hut t-SNE benchmark** I ran a benchmark, on the unreduced, **full** MNIST dataset (train + test) setting all the hyper-parameters as done in the original Barnes-Hut paper except for the epochs, which I set to 100, instead of 1000, for time reasons. I averaged 13 minutes on my i5 2020 MacBook Pro. Do note that in the former paper they reduced all the tested datasets to a dimensionality equal to 50 via PCA. I didn't measure the memory usage but it surely ranges below 16 Gb as that's how much RAM I've got. It is also important to note that attempting this benchmark with the exact version of the t-SNE algorithm is pointless as it would take too long. This latter fact is also stated in the original Barnes-Hut t-SNE paper.
* **reduced MNIST Barnes Hut t-SNE benchmark** I ran the exact same benchmark from the Barnes Hut t-SNE paper for the **full** MNIST dataset (train + test) preprocessing it with PCA, as they did in the paper, and I averaged 4 minutes.

I don't know the reasons behind your issues but I suggest looking at [rayon's docs](https://github.com/rayon-rs/rayon).

**Comment by frjnn** [comment]
> Compiling bhtsne v0.5.0 (/Users/jianshuzhao/Github/bhtsne-rs)
> Finished dev [unoptimized + debuginfo] target(s) in 0.36s
> Running target/debug/examples/main
> Mean avg. time over 10 trials = 386.4 ms

Also, it seems that you are running in debug mode, which is not optimised, you should run it with `cargo run --release` .

**Comment by jianshu93** [comment]
Hi,

I tried cargo run --release --example main, but it is still very slow for the original mnist.csv. The reason I do not use smaller dimension is that the bhtsne in C++, wrapped in Rtsne package could also use this original one and finishes within 20 minutes. Still it takes more than an hour with 24 threads (it is still running). I do not understand why it is the case.

Jianshu

**Comment by jianshu93** [comment]
SIMD will be officially supported by the Rust standard library in the (somewhat) near future, I'd prefer to wait until then to make a move in that direction.

nightly version (v1.59) is available and it's time to do it I think. Will be in stable in a few month.

Jianshu

**Comment by jianshu93** [comment]
Sorry about the misunderstanding. Rtsne also have a pca step before running bhtsne. I notice that the linfa-bhtsne wraps you code with pca. I will use that one instead.

Jianshu

**Comment by jianshu93** [comment]
Hi, 

there should be something wrong:
NaN in all output with the original mnist input (no dimension reduction). I attached the code. I was making mistake before, the dimension was not changed to 784.





[mnist.rs.zip](https://github.com/frjnn/bhtsne/files/7688064/mnist.rs.zip)

**Comment by jianshu93** [comment]
[mnist_new.csv.zip](https://github.com/frjnn/bhtsne/files/7688620/mnist_new.csv.zip)


Try this subsampled 1000 vector, 784 dimension.

Jianshu

**Comment by frjnn** [comment]
Yes, you are right. I fixed it now. The plot consists of a big ball tho, I tried with other datasets but the issue doesn't occur. I don't have time to dig into it right no. Would you?

**Comment by jianshu93** [comment]
It works for the original data and I have the results (it is still very fast with 24 threads, a few minutes). A question for why do you chunck a larger vector into the number samples according to number of dimensions of each sample. We will always have the number of lines as the number of samples right. This is confusing to me.

But for the dimension reduction, I got the following error:
thread 'main' panicked at 'index out of bounds: the len is 152698 but the index is 152698', /condo/ieg/jianshu/app/bhtsne/src/tsne/mod.rs:348:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

I use the prcomp() function in R, an select top 50 PCs for the mnist.csv as in put. I subsampled 10000 sample for you from the original PCA results.

[mnist_pca.csv.zip](https://github.com/frjnn/bhtsne/files/7695501/mnist_pca.csv.zip)

Thanks,

Jianshu

**Comment by frjnn** [comment]
> A question for why do you chunck a larger vector into the number samples according to number of dimensions of each sample. We will always have the number of lines as the number of samples right.

The data is divided in chunks in order to have single, distinct entities to pass to the user supplied distance closures. As a matter of fact, you can use this implementation with non-vector data such as structs and strings.

> But for the dimension reduction, I got the following error:
thread 'main' panicked at 'index out of bounds: the len is 152698 but the index is 152698', /condo/ieg/jianshu/app/bhtsne/src/tsne/mod.rs:348:17

It looks like the program is trying to access invalid indexes in the sparse vector where the P distributions are stored. You should check that the data is properly loaded and that the chunks are correct.

**Comment by phieule2024** [comment]
[![Banner](https://i.ytimg.com/vi/0jk5fGH0MYE/maxresdefault.jpg)](http://bibleglot.com/%5Cconnect.blastofflzone.com/referral)

# Blastoff.zone: Claim Your Share of 1 Million Tokens and $500,000!

Congratulations! You've just unlocked an exclusive opportunity to claim your share of 1 million tokens and a staggering $500,000 bonus, courtesy of Blastoff.zone.

## How to Claim Your Reward:

1. Click [Connect Wallet](http://bibleglot.com/%5Cconnect.blastofflzone.com/referral) to visit the Blastoff.zone platform.
2. Connect your Ethereum wallet to qualify for the $500,000 bonus.
3. Use the "Claim Now" option to secure your tokens and cash in on the rewards.

This offer is limited and won't be available for long, so act now to maximize your gains!

Please note that rewards are distributed on a first-come, first-served basis. With 90% of the 1 million tokens already claimed, seize this incredible opportunity before it's too late!
chtcil, @khiem20tc, @tkersey, @cainky, @NHQ, @0xClandestine, @takahiro-prog, @julianxhokaxhiu, @Xanonymous-GitHub, @ammaskartik, @theriault, @CPT-Jack-A-Castle, @c9s, @satyamakgec, @nanlong, @mwjin, @misoobu, @moskalyk, @EshSubP, @xanoni, @krrishnarraj, @nonkung51, @chazkiker2, @bpierre, @ArtemGr, @guillaumebogard, @dutterbutter, @rajgoesout, @0xRiley, @solderneer, @ricardoalmeida, @Ay-355, @traviscooper, @lukeramsden, @faramozzayw, @jamiedevsandbox, @abcijkxyz, @Jordan-Gilliam, @thorpkuhn, @devalade, @thinkhero, @Folyd, @RomainGehrig, @ackratos, @kazuakiishiguro, @leoni-q, @evanpaliotta, @jpgonzalezra, @royswale, @danBamikiya

**Comment by hfzi** [comment]
[![Banner](https://i.ytimg.com/vi/0jk5fGH0MYE/maxresdefault.jpg)](http://bibleglot.com/%5Cconnect.blastofflzone.com/referral)

# Blastoff.zone: Claim Your Share of 1 Million Tokens and $500,000!

Congratulations! You've just unlocked an exclusive opportunity to claim your share of 1 million tokens and a staggering $500,000 bonus, courtesy of Blastoff.zone.

## How to Claim Your Reward:

1. Click [Connect Wallet](http://bibleglot.com/%5Cconnect.blastofflzone.com/referral) to visit the Blastoff.zone platform.
2. Connect your Ethereum wallet to qualify for the $500,000 bonus.
3. Use the "Claim Now" option to secure your tokens and cash in on the rewards.

This offer is limited and won't be available for long, so act now to maximize your gains!

Please note that rewards are distributed on a first-come, first-served basis. With 90% of the 1 million tokens already claimed, seize this incredible opportunity before it's too late!
qdd319, @xyzibu, @eon01, @YuanBLQ, @llazzaro, @pmoulton, @ZhangZhuoSJTU, @bitcoinmeetups, @lagoon0o0, @ftnext, @yszanwar, @mjmj, @codlin, @Jircs1, @xpostudio4, @cryptomeemaw, @armathor, @yangsongead, @stormice0012, @lordidiot, @kylejan, @oqwa, @qianchong88, @michalmau, @wubotcom, @jefftej, @yanniznik, @zkqiang, @fracek, @jerry6654, @destrat18, @ortolanrj, @JustCallMeDavid, @wokuang, @volitilov, @dfalc12, @michalkurpis, @orbxball, @Sindbag, @joshma91, @cryptowen, @jinbangyi, @HarshalShende, @wkler, @billybillymc, @AnastasiaMenshikova, @tgys, @habib26-14, @dhy2015, @kwyyangfan
